From a00d0e141973bf69c4f566962d05d6204111748c Mon Sep 17 00:00:00 2001
From: Wang Hui <Hui.Wang@windriver.com>
Date: Wed, 16 Jan 2013 09:47:08 +0800
Subject: [PATCH 3/5] arm: armadaxp: errata: add workaround for Sheeva errata 6043

Sheeva Errata 6043: clean operations can cause victim data to be
written out of order.

The code is extracted from linux-3.2.27-axp_a370-2012_Q4.1, which
can be downloaded from:
https://extranet.marvell.com/extranet/dms/documents.do?groupID=4&\
subGroupID=53015

Signed-off-by: Wang Hui <Hui.Wang@windriver.com>
---
 arch/arm/include/asm/tlbflush.h |   10 ++++++++++
 arch/arm/mm/cache-v7.S          |    8 ++++++++
 2 files changed, 18 insertions(+), 0 deletions(-)

diff --git a/arch/arm/include/asm/tlbflush.h b/arch/arm/include/asm/tlbflush.h
index 6bb71e9..75fc373 100644
--- a/arch/arm/include/asm/tlbflush.h
+++ b/arch/arm/include/asm/tlbflush.h
@@ -475,7 +475,12 @@ static inline void flush_pmd_entry(void *pmd)
 		raw_local_irq_save(flags);
 		dmb();
 #endif
+#if defined(CONFIG_SHEEVA_ERRATA_ARM_CPU_6043) || defined(CONFIG_SHEEVA_ERRATA_ARM_CPU_6124)
+		asm("mcr	p15, 0, %0, c7, c14, 1  @ flush_pmd"
+			: : "r" (pmd) : "cc");
+#else
 	tlb_op(TLB_DCLEAN, "c7, c10, 1	@ flush_pmd", pmd);
+#endif
 #ifdef CONFIG_SHEEVA_ERRATA_ARM_CPU_4611
 		raw_local_irq_restore(flags);
 	}
@@ -503,7 +508,12 @@ static inline void clean_pmd_entry(void *pmd)
                 raw_local_irq_save(flags);
                 dmb();
 #endif
+#if defined(CONFIG_SHEEVA_ERRATA_ARM_CPU_6043) || defined(CONFIG_SHEEVA_ERRATA_ARM_CPU_6124)
+		asm("mcr	p15, 0, %0, c7, c14, 1  @ flush_pmd"
+			: : "r" (pmd) : "cc");
+#else
 	tlb_op(TLB_DCLEAN, "c7, c10, 1	@ flush_pmd", pmd);
+#endif
 #ifdef CONFIG_SHEEVA_ERRATA_ARM_CPU_4611
 		raw_local_irq_restore(flags);
 	}
diff --git a/arch/arm/mm/cache-v7.S b/arch/arm/mm/cache-v7.S
index 2e280fb..8df01c6 100644
--- a/arch/arm/mm/cache-v7.S
+++ b/arch/arm/mm/cache-v7.S
@@ -193,7 +193,11 @@ ENTRY(v7_coherent_user_range)
 	ALT_UP(W(nop))
 #endif
 1:
+#if defined CONFIG_SHEEVA_ERRATA_ARM_CPU_6043 || defined(CONFIG_SHEEVA_ERRATA_ARM_CPU_6124) || defined(CONFIG_SHEEVA_ERRATA_ARM_CPU_6076)
+	USER(  mcr     p15, 0, r12, c7, c14, 1 )       @ clean & invalidate D line to the point of unification
+#else
  USER(	mcr	p15, 0, r12, c7, c11, 1	)	@ clean D line to the point of unification
+#endif
 	add	r12, r12, r2
 	cmp	r12, r1
 	blo	1b
@@ -304,7 +308,11 @@ v7_dma_clean_range:
 	ALT_UP(W(nop))
 #endif
 1:
+#if defined CONFIG_SHEEVA_ERRATA_ARM_CPU_6043 || defined(CONFIG_SHEEVA_ERRATA_ARM_CPU_6124) || defined(CONFIG_SHEEVA_ERRATA_ARM_CPU_6076)
+	mcr		p15, 0, r12, c7, c14, 1		@ clean & invalidate D line to the point of unification
+#else
 	mcr	p15, 0, r0, c7, c10, 1		@ clean D / U line
+#endif
 	add	r0, r0, r2
 	cmp	r0, r1
 	blo	1b
-- 
1.7.0.2

